feat: add `stats/strided/dminabssorted`

DivyanshuVortex opened the pull request:
---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes. report:
  - task: lint_filenames status: passed
  - task: lint_editorconfig status: passed
  - task: lint_markdown status: passed
  - task: lint_package_json status: passed
  - task: lint_repl_help status: passed
  - task: lint_javascript_src status: passed
  - task: lint_javascript_cli status: na
  - task: lint_javascript_examples status: passed
  - task: lint_javascript_tests status: passed
  - task: lint_javascript_benchmarks status: na
  - task: lint_python status: na
  - task: lint_r status: na
  - task: lint_c_src status: na
  - task: lint_c_examples status: na
  - task: lint_c_benchmarks status: na
  - task: lint_c_tests_fixtures status: na
  - task: lint_shell status: na
  - task: lint_typescript_declarations status: passed
  - task: lint_typescript_tests status: passed
  - task: lint_license_headers status: passed ---

Resolves #{{TODO: add issue number}}.

## Description

> What is the purpose of this pull request?

This pull request:

-   Adds `stats/strided/dminabssorted`

## Related Issues

> Does this pull request have any related issues?

This pull request has the following related issues:

-  none

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

### AI Assistance

> When authoring the changes proposed in this PR, did you use any kind of AI assistance?

-   [ ] Yes
-   [x] No

If you answered "yes" above, how did you use AI assistance?

-   [ ] Code generation (e.g., when writing an implementation or fixing a bug)
-   [ ] Test/benchmark generation
-   [ ] Documentation (including examples)
-   [ ] Research and understanding

#### Disclosure


{{TODO: add disclosure if applicable}}

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

DivyanshuVortex committed:
feat:add stats/strided/dminabssorted

DivyanshuVortex committed:
feat(stats/strided): add dminabssorted

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: passed
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

DivyanshuVortex committed:
feat(stats/strided): add sminabssorted fix

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: missing_dependencies
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

DivyanshuVortex committed:
feat(stats/strided): add sminabssorted fix

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: missing_dependencies
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

DivyanshuVortex committed:
feat(stats/strided): add dminabssorted fix benchmark

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: missing_dependencies
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

DivyanshuVortex committed:
feat(stats/strided): add dminabssorted fix benchmark

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

kgryte commented on lib/node_modules/@stdlib/stats/strided/dminabssorted/benchmark/c/benchmark.length.c:
```diff
@@ -0,0 +1,192 @@
+	count = 0;
+	for ( i = MIN; i <= MAX; i++ ) {
+		len = pow( 10, i );
+		iter = ITERATIONS;
```
```suggestion
		iter = ITERATIONS / pow( 10, i-1 );
```

Not clear why this line was changed, but this is likely the reason for the benchmarks timing out.

kgryte committed:
bench: fix iterations

Signed-off-by: Athan <kgryte@gmail.com>

DivyanshuVortex commented:
Hey @kgryte , pls review my code

Planeshifter commented on lib/node_modules/@stdlib/stats/strided/dminabssorted/test/test.ndarray.js:
```diff
@@ -0,0 +1,189 @@
+	v = dminabssorted( x.length, x, 1, 0 );
+	t.strictEqual( v, 1.0, 'returns expected value' );
+
+	x = new Float64Array( [ 3.0, 2.0, 1, 0.0, -1.0, -2.0, -3.0 ] );
```
Inconsistent float literal style. Float64Array values should use `.0` suffix for consistency (e.g., `1.0` not `1`).

```suggestion
	x = new Float64Array( [ 3.0, 2.0, 1.0, 0.0, -1.0, -2.0, -3.0 ] );
```

Planeshifter commented on lib/node_modules/@stdlib/stats/strided/dminabssorted/test/test.ndarray.native.js:
```diff
@@ -0,0 +1,198 @@
+	v = dminabssorted( x.length, x, 1, 0 );
+	t.strictEqual( v, 1.0, 'returns expected value' );
+
+	x = new Float64Array( [ 3.0, 2.0, 1, 0.0, -1.0, -2.0, -3.0 ] );
```
Inconsistent float literal style. Float64Array values should use `.0` suffix for consistency (e.g., `1.0` not `1`).

```suggestion
	x = new Float64Array( [ 3.0, 2.0, 1.0, 0.0, -1.0, -2.0, -3.0 ] );
```

Planeshifter reviewed (commented):
Thanks for opening this PR!

DivyanshuVortex committed:
fix(dminabssorted): fix package stats/strided/dminabssorted

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

Planeshifter commented:
/stdlib update-copyright-years

DivyanshuVortex committed:
fix(dminabssorted): fix package stats/strided/dminabssorted

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: passed
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: missing_dependencies
  - task: lint_c_examples
    status: missing_dependencies
  - task: lint_c_benchmarks
    status: missing_dependencies
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: passed
  - task: lint_license_headers
    status: passed
---

Planeshifter commented on lib/node_modules/@stdlib/stats/strided/dminabssorted/README.md:
```diff
@@ -0,0 +1,305 @@
+var Float64Array = require( '@stdlib/array/float64' );
+
+var x = new Float64Array( [ -1.0, -2.0, -3.0 ] );
+var v = dminabssorted( x.length, x, 1 );
```
For consistency with `dminsorted` and other strided packages, add a blank line between the variable declaration and function call. The reference package has:

```javascript
var x = new Float64Array( [ 1.0, 2.0, 3.0 ] );

var v = dminsorted( x.length, x, 1 );
```

```suggestion
var x = new Float64Array( [ -1.0, -2.0, -3.0 ] );

var v = dminabssorted( x.length, x, 1 );
```

Planeshifter commented on lib/node_modules/@stdlib/stats/strided/dminabssorted/README.md:
```diff
@@ -0,0 +1,305 @@
+var Float64Array = require( '@stdlib/array/float64' );
+
+var x = new Float64Array( [ -1.0, -2.0, -3.0 ] );
+var v = dminabssorted.ndarray( x.length, x, 1, 0 );
```
Same pattern here - add blank line between variable declaration and function call:

```suggestion
var x = new Float64Array( [ -1.0, -2.0, -3.0 ] );

var v = dminabssorted.ndarray( x.length, x, 1, 0 );
```

kgryte commented:
/stdlib merge

stdlib-bot committed:
Merge remote-tracking branch 'upstream/develop' into stats/strided/dmidabssorted

DivyanshuVortex committed:
style: add structure to benchmark

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

Planeshifter commented on lib/node_modules/@stdlib/stats/strided/dminabssorted/src/main.c:
```diff
@@ -0,0 +1,82 @@
+		if ( v1 < minAbs ) {
+			minAbs = v1;
+		}
+}
```
The closing brace for the `for` loop is missing its tab indentation. It should align with the `for` statement on line 74.

```suggestion
	}
```

DivyanshuVortex committed:
style: add structure to src c file

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: missing_dependencies
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

kgryte commented on lib/node_modules/@stdlib/stats/strided/dminabssorted/README.md:
```diff
@@ -0,0 +1,305 @@
+
+-   <span class="package-name">[`@stdlib/stats/strided/dminabs`][@stdlib/stats/strided/dminabs]</span><span class="delimiter">: </span><span class="description">calculate the minimum absolute value of a double-precision floating-point strided array.</span>
+-   <span class="package-name">[`@stdlib/stats/strided/dminsorted`][@stdlib/stats/strided/dminsorted]</span><span class="delimiter">: </span><span class="description">calculate the minimum value of a sorted double-precision floating-point strided array.</span>
+-   <span class="package-name">[`@stdlib/stats/strided/sminabssorted`][@stdlib/stats/strided/sminabssorted]</span><span class="delimiter">: </span><span class="description">calculate the minimum absolute value of a sorted single-precision floating-point strided array.</span>
```
```suggestion
```

New packages should not have any content in this section.

kgryte commented on lib/node_modules/@stdlib/stats/strided/dminabssorted/README.md:
```diff
@@ -0,0 +1,305 @@
+[@stdlib/stats/strided/sminabssorted]: https://github.com/stdlib-js/stdlib/tree/develop/lib/node_modules/%40stdlib/stats/strided/sminabssorted
+
+<!-- </related-links> -->
+
```
```suggestion
```

kgryte commented on lib/node_modules/@stdlib/stats/strided/dminabssorted/lib/ndarray.js:
```diff
@@ -0,0 +1,85 @@
+	}
+	minAbs = abs( v1 );
+	for ( i = 1; i < N; i++ ) {
+		val = abs( x[ offsetX + ( i * strideX ) ] );
```
This is not desired. Please see other packages. For every loop iteration, you are performing two arithmetic operations. There is a reason why other packages use `ix += strideX`. Follow them.

kgryte replied:
This still needs to be updated.

kgryte commented on lib/node_modules/@stdlib/stats/strided/dminabssorted/lib/ndarray.js:
```diff
@@ -0,0 +1,85 @@
+		return v2;
+	}
+	minAbs = abs( v1 );
+	for ( i = 1; i < N; i++ ) {
```
But also, why is this loop even here? When do we need to iterate over the input array? The array is supposed to be sorted.

This package should have been a simple copy-paste of https://github.com/stdlib-js/stdlib/blob/develop/lib/node_modules/%40stdlib/stats/strided/dmaxabssorted with minor modifications. If you attempted to use AI to do this, you are vastly overcomplicating what should be a trivial contribution.

DivyanshuVortex replied:
@kgryte sir , I believe `dminabssorted` cannot be implemented in O(1) in the general case and must perform an O(N) linear scan and `dmaxabssorted` works in O(1) because, for a sorted array, the maximum absolute value is always at one of the ends.

like : 
ip ->  -7 -2 4 8 10 
steps - 
abs(-7) === 7
abs(10) === 10 
10 > 7 
so return 10 [ correct ] . 
However, the minimum absolute value is not guaranteed to be at the boundaries when the array crosses zero:

ip -> -7 -2 4 8 10
steps - 
abs(-7) === 7
abs(-10) === 10
7 < 10
returns 7 { wrong bcoz 2 is the absolute minimum }

Therefore , I added a fallback linear scan for dminabssorted when the endpoints have opposite signs.
It maybe an option to implemente it  by the Binary search but its change the whole strucutre of the files too much . Please let me know if comparing only the first, last, and zero is considered sufficient here, or am I thinking in a totally wrong direction ?


Should `dminabssorted` :

Follow the same endpoint-only convention as other *sorted kernels (even if that does not always return the true minimum absolute value), or

*Guarantee* the true minimum absolute value, which would require scanning or a more complex search?

I want to confirm which behavior is expected before proceeding **

kgryte replied:
You're right. I stand corrected. We cannot follow the same approach as `dmaxabssorted`. Will take another look.

DivyanshuVortex committed:
fix: removing the ref and related link in readme

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

DivyanshuVortex committed:
fix: removing the two operation in dminabssorted

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: missing_dependencies
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

DivyanshuVortex committed:
Merge branch 'stdlib-js:develop' into stats/strided/dmidabssorted

DivyanshuVortex committed:
Merge branch 'develop' into stats/strided/dmidabssorted
